### PR TITLE
doctor: make JDK validation message more descriptive

### LINKED ISSUE #153156: `flutter doctor -v ` should tell what setting has set the JDK source.
I suddenly came upon this error message when building Flutter after upgrading my Android Studio to v2024.2.1.1 (LadyBug Canary 5):

```
           ┌─ Flutter Fix ───────────────────────────────────────────────────────────────────────────────────┐
           │ [!] Your project's Gradle version is incompatible with the Java version that Flutter is using   │
           │ for Gradle.                                                                                     │
           │                                                                                                 │
           │ If you recently upgraded Android Studio, consult the migration guide at                         │
           │ https://flutter.dev/to/to/java-gradle-incompatibility.                                          │
           │                                                                                                 │
           │ Otherwise, to fix this issue, first, check the Java version used by Flutter by running `flutter │
           │ doctor --verbose`.                                                                              │
           │                                                                                                 │
           │ Then, update the Gradle version specified in                                                    │
           │ C:\FlutterProjects\mwork\android\gradle\wrapper\gradle-wrapper.properties to be compatible with │
           │ that Java version. See the link below for more information on compatible Java/Gradle versions:  │
           │ https://docs.gradle.org/current/userguide/compatibility.html#java                               │
           │                                                                                                 │
           │                                                                                                 │
           └─────────────────────────────────────────────────────────────────────────────────────────────────┘

```
I work in different Android Studio versions, in Flutter I am in AS 2024.1.1 Koala Patch 2, but for my kotlin projects I am in AS 2024.2.1.1  (LadyBug Canary 5). 

It seems that suddenly my Flutter (as FLutter Doctor says) uses java 21, which sadly isn't compatible with Flutter 3.24.0: 
```
[√] Android toolchain - develop for Android devices (Android SDK version 35.0.0)
    • Android SDK at C:\Users\roarg\AppData\Local\Android\Sdk
    • Platform android-35, build-tools 35.0.0
    • ANDROID_HOME = C:\Users\roarg\AppData\Local\Android\Sdk
    • Java binary at: C:\Program Files\Android Studio\Android Studio 2024.2.1.1\jbr\bin\java
    • Java version OpenJDK Runtime Environment (build 21.0.3+-12099254-b509.4)
    • All Android licenses accepted.

```

My next problem: I cannot see from `flutter doctor -v` where Flutter knows what java version to use. Is it the largest/last version installed or found ?

You should give information on HOW Flutter has overcome this java version. 

And you should give information on how to set the version you want Flutter to use.

### PR DESCRIPTION
**Pull request by Sameri11**
This PR attempts to improve clarity of androids section of  `flutter doctor -v` output by providing explicit information about which JDK is being used and how to configure a different one if needed.

### Before

```console
• Java binary at: /Users/user/Applications/Android Studio Ladybug Feature Drop 2024.2.2 Canary 2.app/Contents/jbr/Contents/Home/bin/java
```

### After

1. When JDK is from Android Studio:

```console
    • Java binary at: /Users/users/Applications/Android Studio Ladybug Feature Drop 2024.2.2 Canary 2.app/Contents/jbr/Contents/Home/bin/java
      This is the JDK bundled with latest Android Studio installation
      To manually set a custom JDK path, use: `flutter config --jdk-dir="path/to/jdk"`
```

2. When JDK is from JAVA_HOME env variable:

```console
    • Java binary at: /Users/user/Applications/Android Studio Ladybug Feature Drop 2024.2.2 Canary 2.app/Contents/jbr/Contents/Home/bin/java
      This JDK is specified by JAVA_HOME environment variable
      To manually set a custom JDK path, use: `flutter config --jdk-dir="path/to/jdk"`
```

3. When path to JDK is set in flutter config:

```console
    • Java binary at: /Users/user/Applications/Android Studio Ladybug Feature Drop 2024.2.2 Canary 2.app/Contents/jbr/Contents/Home/bin/java
      This JDK was found in system PATH
      To change current JDK, run: `flutter config --jdk-dir="path/to/jdk"`
```
4. When java binary is found in PATH (as fallback)

```console
    • Java binary at: /Users/user/Applications/Android Studio Ladybug Feature Drop 2024.2.2 Canary 2.app/Contents/jbr/Contents/Home/bin/java
      This JDK is specified in Flutter configuration
      To manually set a custom JDK path, use: `flutter config --jdk-dir="path/to/jdk"`
```

### Motivation

I think it's described in https://github.com/flutter/flutter/issues/153156#issuecomment-2336814991. 

TLDR; many developers struggle with Java-related issues and more verbose doctor's output will (presumably) improve DX in that part.

fixes #153156

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

### CODE REVIEW

**Comment by andrewkolos** [comment]
I pushed commits with small tweaks. Feel free to revise and revert as desired.

**Review by andrewkolos — APPROVED** [review]
LGTM! Thanks for the improvement here :)

**Comment by andrewkolos** [comment]
It looks like I broke tests, I'll work on fixing them

**Comment by Sameri11** [comment]
@andrewkolos thanks for your commit! I guess I missed deprecation of user_messages. Also, please tell if you need any assistance with fixing those tests.

**Comment by andrewkolos** [comment]
friendly ping @bkonyi

**Review by bkonyi — APPROVED** [review]
LGTM!

**Comment by twitchard** [comment]
Thank you for doing this. Had I been using a version of flutter that included this change it would have saved ~2 hours of my time today.
